### PR TITLE
Fix AddAppt / DelAppt incorrect messages

### LINKED ISSUE #307: [Bug] Incorrect error messages when using invalid timings
When varying the length/using the wrong length of the time for `addAppt` and `delAppt`, the error messages shown in the display boxes are not appropriate.

image

image

### LINKED ISSUE #308: [Bug] Incorrect error message for wrong appointment date
Bug:
When entering a date that does not exist, `MESSAGE_CONSTRAINTS_APPT_DATE_WRONG_FORMAT` is shown instead. 
Based on your code in `Appointment` class, `MESSAGE_CONSTRAINTS_APPT_DATE_INVALID_DATE_1S` should be shown instead

image

Possible fix:
According to how `Appointment` class is checking arguments, `AddApptCommandParser` and `DelApptCommandParser` should have another conditional check using `isCorrectDateFormat` where `MESSAGE_CONSTRAINTS_APPT_DATE_WRONG_FORMAT` will be shown instead before `isValidDate`

For `isValidDate` condition, the message shown should be `MESSAGE_CONSTRAINTS_APPT_DATE_INVALID_DATE_1S`

image

### LINKED ISSUE #322: [Bug] Incorrect error messages when adding appointment that clashes
image

Error message suggests duplicate appointment when it is a clashing event instead.

### PR DESCRIPTION
**Pull request by yadobler**
fixes #307
fixes #308
fixes #322

### CODE REVIEW

**Review by yong-zhuo — APPROVED** [review]
LGTM, fixes are appropriate for the error messages and test cases are designed well

File: src/main/java/seedu/address/logic/commands/AddApptCommand.java [hunk 1]
```diff
@@ -130,8 +131,10 @@ private static Person createEditedPerson(Person personToEdit,
         try {
             oldAppointmentList = new ArrayList<>(personToEdit.getAppointments());
             for (Appointment a : oldAppointmentList) {
-                if (a.isClashing(newApptDate, newApptTime)) {
+                if (a.isSameDateTime(newApptDate, newApptTime)) {
                     throw new CommandException(String.format(MESSAGE_DUPLICATE_APPT_1S, a));
+                } else if (a.isClashing(newApptDate, newApptTime)) {
+                    throw new CommandException(String.format(MESSAGE_CLASHING_APPT_1S, a));
```
Comment (TechRaj) [inline src/main/java/seedu/address/logic/commands/AddApptCommand.java:137]: good catch

File: src/main/java/seedu/address/model/person/AppointmentExistsPredicate.java [hunk 2]
```diff
@@ -28,12 +29,26 @@ public AppointmentExistsPredicate(String date, String timePeriod) {
...
+     * Wraps throwable boolean method {@link Appointment#isSameDateTime} to be used in stream.
+     *
+     * @param x Appointment to be tested.
+     * @return result of <code>isSameDateTime</code> or false if exception thrown since invalid input never matches.
+     */
+    private boolean safeDateTimeTest(Appointment x) {
+        try {
+            return x.isSameDateTime(date, timePeriod);
+        } catch (IllegalValueException e) {
+            return false;
```
Comment (TechRaj) [inline src/main/java/seedu/address/model/person/AppointmentExistsPredicate.java:43]: good idea to abstract this out

File: src/test/java/seedu/address/logic/commands/AddApptCommandTest.java [hunk 3]
```diff
@@ -118,6 +122,43 @@ public void execute_patientDuplicateAppointments_failure() {
...
+
+        AddApptCommand addApptPersonCommand3 = new AddApptCommand(nricAmy, VALID_APPOINTMENT_NAME_AMY,
+                                                                  VALID_APPOINTMENT_DATE_AMY,
+                                                                  INVALID_APPOINTMENT_TIMEPERIOD_CLASH_DENTAL_3);
+        assertCommandFailure(addApptPersonCommand3, model, expectedErrorMessage);
+
+        AddApptCommand addApptPersonCommand4 = new AddApptCommand(nricAmy, VALID_APPOINTMENT_NAME_AMY,
+                                                                  VALID_APPOINTMENT_DATE_AMY,
+                                                                  INVALID_APPOINTMENT_TIMEPERIOD_CLASH_DENTAL_4);
+        assertCommandFailure(addApptPersonCommand4, model, expectedErrorMessage);
```
Comment (TechRaj) [inline src/test/java/seedu/address/logic/commands/AddApptCommandTest.java:159]: Thorough testing!

**Review by TechRaj — APPROVED** [review]
LGTM

**Pull request merged**